### PR TITLE
feat: guidance playground

### PR DESCRIPTION
**Pull request by danielsanchezaran**
# Guidance Framework & Interactive GUI

## Summary

This PR adds an interactive guidance exploration tool (`guidance_gui/`) and the underlying
guidance framework that powers it, built on top of the existing DPO annotation pipeline.

### Guidance Framework (`diffusion_planner/model/guidance/`)

A structured framework for composing guidance functions during DPM-Solver sampling:

- **`BaseGuidance`** — abstract base class; each subclass implements `_compute(x, inputs) → [B]`
- **`GuidanceConfig` / `GuidanceSetConfig`** — dataclasses describing which functions are active,
  their per-function scales, and the global scale multiplier
- **`GuidanceComposer`** — drop-in replacement for `GuidanceWrapper`; builds from a
  `GuidanceSetConfig` and applies all active functions
- **`@register` / `build()` / `list_available()`** — registry for guidance functions; new
  functions are added by decorating their class, with no changes elsewhere
- Existing guidance functions (`collision`, `route_following`, `lane_keeping`,
  `centerline_following`) refactored into registered classes with backward-compatible
  module-level aliases
- **`AnchorFollowingGuidance`** — new MTR-style guidance that attracts the ego trajectory
  toward a pre-clustered prototype trajectory shape. Prototype file is loaded once per path
  via `lru_cache` to avoid repeated disk I/O on every UI interaction

### Guidance GUI (`guidance_gui/`)

A Gradio web app for interactively exploring guidance on arbitrary dataset samples:

- Generates N independent trajectory samples under configurable noise + guidance
- Per-function enable/scale sliders; global scale slider
- Prototype gallery: 16 KMeans-clustered trajectory shapes from the training set,
  click to select as anchor for `AnchorFollowingGuidance`
- Navigation controls (prev/next/jump), velocity and lateral curvature plots, sample stats
- `python3 -m guidance_gui --model_path <ckpt> --npz_list <json>` — auto-opens browser

### Shared guidance UI panel (`guidance_gui/guidance_ui.py`)

`build_guidance_panel()` and `make_guidance_set_config()` are shared between the Guidance GUI
and the DPO annotation GUI. Adding a new guidance function requires changes in one place only.

### DPO annotation GUI changes (`preference_optimization/annotation_gui.py`)

- Guidance controls added to the annotation UI using the shared panel
- Guided trajectory (`traj_2`) is generated with `noise_scale ≈ 0` to cleanly isolate the
  guidance signal from sampling noise (see `docs/guidance_gui_design.md` §10 for rationale)
- Stochastic trajectory auto-regenerates when any guidance control changes

---

## Files Changed

| Path | Change |
|---|---|
| `diffusion_planner/model/guidance/base.py` | NEW — `BaseGuidance` abstract class |
| `diffusion_planner/model/guidance/config.py` | NEW — `GuidanceConfig`, `GuidanceSetConfig` |
| `diffusion_planner/model/guidance/registry.py` | NEW — `@register`, `build()`, `list_available()` |
| `diffusion_planner/model/guidance/composer.py` | NEW — `GuidanceComposer` |
| `diffusion_planner/model/guidance/anchor_following.py` | NEW — `AnchorFollowingGuidance` |
| `diffusion_planner/model/guidance/{collision,route_following,lane_keeping,centerline_following}.py` | MODIFIED — refactored to classes, backward-compat aliases kept |
| `diffusion_planner/model/guidance/__init__.py` | MODIFIED — exports framework |
| `guidance_gui/` | NEW — Gradio exploration app, shared UI panel, visualisation utilities |
| `guidance_gui/prototypes_k16.npy` | NEW — 16 prototype trajectory shapes (Shinagawa-Odaiba dataset) |
| `guidance_gui/scripts/generate_prototypes.py` | NEW — `scipy.cluster.vq.kmeans2` clustering script |
| `preference_optimization/annotation_gui.py` | MODIFIED — guidance controls added |
| `preference_optimization/utils.py` | MODIFIED — migrated to `GuidanceSetConfig` API; deprecated boolean flags kept with warning |
| `docs/guidance_gui_design.md` | NEW (moved from `guidance_playground/DESIGN.md`) |
| `docs/guidance_framework_design.md` | NEW (moved from `guidance_playground/GUIDANCE_FRAMEWORK_DESIGN.md`) |

---

## Design Decisions

**Scale convention:** each guidance class has an `_energy_scale` class attribute (the original
internal constant). `_compute()` returns the raw reward without it applied. `energy()` =
`_energy_scale × config.scale × _compute()`. A slider value of `1.0` reproduces the original
behavior; `2.0` doubles the strength.

**DPO GUI uses `noise_scale = 0` for the guided trajectory** to avoid conflating guidance
signal with noise-induced artifacts. High noise causes erratic first waypoints; if those are
then used as "preferred" examples in DPO, the model learns to replicate the erratic behavior.
The unguided baseline stays stochastic for diversity.

**Anchor prototypes are dataset-specific.** The committed `prototypes_k16.npy` was generated
from the Shinagawa-Odaiba training set. To regenerate for a different dataset:
```bash
python guidance_gui/scripts/generate_prototypes.py \
  --npz_list /path/to/train.json --k 16 --output guidance_gui/prototypes_k16.npy
```

---

## Testing

```bash
# Import checks
source .venv/bin/activate
python -c "from guidance_gui.guidance_ui import build_guidance_panel, make_guidance_set_config; print('OK')"
python -c "from preference_optimization.annotation_gui import PreferenceAnnotator; print('OK')"

# Smoke test
python -c "
from guidance_gui.guidance_ui import make_guidance_set_config
cfg = make_guidance_set_config(True, True, 1.0, False, 1.0, False, 1.0, False, 1.0, False, 1.0, 0, '', 0.5)
print('active:', [f.name for f in cfg.functions if f.enabled])
"

# Launch GUI
python3 -m guidance_gui \
  --model_path /path/to/model.pth \
  --npz_list /path/to/dpo-npz.json \
  --prototypes guidance_gui/prototypes_k16.npy
```

### CODE REVIEW

File: guidance_gui/scripts/generate_prototypes.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import argparse
+import json
+import os
+import random
+
+import numpy as np
+from scipy.cluster.vq import kmeans2, whiten
+
+
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply guidance_gui/scripts/generate_prototypes.py:27]: Resolved it in the docs rather than the script — the design doc was wrong, not the  implementation. scipy is already a project dependency (listed in requirements.txt), while scikit-learn would be an unnecessary addition. The suggested replacement also has subtle behavioral differences: scipy.cluster.vq.whiten normalises by the RMS (column  norm), not the standard deviation, and kmeans2 with minit="points" and iter=20 gives better convergence guarantees than   KMeans(n_init=10) with a fixed random_state. The docs and implementation are now aligned on scipy.cluster.vq.

File: diffusion_planner/diffusion_planner/model/guidance/anchor_following.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    """
+
+    name = "anchor_following"
+    _energy_scale = 0.05
+
+    def __init__(self, config: "GuidanceConfig"):  # noqa: F821
+        super().__init__(config)
+        protos = np.load(config.params["prototypes_path"])  # (K, 80, 2)
+        idx = config.params["anchor_index"]
+        self._anchor = torch.tensor(protos[idx], dtype=torch.float32)  # (80, 2)
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply diffusion_planner/diffusion_planner/model/guidance/anchor_following.py:47]: Already fixed in this PR — added a module-level _load_prototypes(path) function decorated with @functools.lru_cache(maxsize=4) so repeated construction of AnchorFollowingGuidance with the same path hits the in-memory cache rather than disk.

File: guidance_gui/guidance_ui.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+    # Wire gallery events once here so callers never have to repeat this boilerplate.
+    # Gradio 6.x requires gr.SelectData type annotation for SelectData to be injected;
+    # bare lambdas receive None without it.
+    def _on_gallery_select(evt: gr.SelectData, path: str) -> tuple:
+        updated = render_prototype_gallery(path, selected_index=evt.index) or []
+        return (
+            gr.update(value=evt.index, maximum=max(63, evt.index)),
+            gr.update(value=updated),
+        )
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply guidance_gui/guidance_ui.py:243]: fixed

File: guidance_gui/guidance_ui.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    with gr.Accordion("Prototype Gallery — click to select anchor", open=False):
+        _default_gallery = render_prototype_gallery(default_prototypes_path)
+        gallery = gr.Gallery(
+            value=_default_gallery or [],
+            columns=8,
+            rows=2,
+            height=260,
+            allow_preview=False,
+            selected_index=0,
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply guidance_gui/guidance_ui.py]: fixed in [fd7cd13](https://github.com/tier4/Diffusion-Planner/pull/22/commits/fd7cd13a73f158861605f5591b895b1f2d8339e6)

File: diffusion_planner/diffusion_planner/model/guidance/base.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            x: [B, P, T+1, 4] trajectory in physical ego-centric metres,
+               x_start-corrected and state_normalizer.inverse applied by
+               GuidanceComposer before this call.
+            t: [B] diffusion timestep in [0, 1].
+            inputs: Observation dict in physical units.
+
+        Returns:
+            [B] energy tensor.
+        """
+        mask = (t < 0.1) * (t > 0.005)
```
Comment (SakodaShintaro) [inline diffusion_planner/diffusion_planner/model/guidance/base.py]: Should this be set in `Base` class?
What is the background of these values (0.1, 0.005)?
If we use 10 steps for denoising, guidance will be conducted at only one last step?

Comment (danielsanchezaran) [reply diffusion_planner/diffusion_planner/model/guidance/base.py]: @SakodaShintaro The values 0.005 and 0.1 originate from the original upstream guidance implementation and were centralized into BaseGuidance during the refactoring to avoid duplicating the mask in every guidance function. They gate guidance to the near-clean regime of the VPSDE diffusion time axis (t∈[0,1], where t=1 is pure noise and t=0 is clean trajectory):                                               
                                                                                                                                     
  - t > 0.1: the trajectory is still noise-dominated; geometric quantities like collision distances and lane boundaries computed from x are not yet meaningful, so gradients point in unreliable directions.                                                            
  - t < 0.005: the trajectory is essentially clean and the DPM-Solver is making very fine corrections; applying guidance here causes numerical instability.
  
Regarding the step count: with 10 logSNR-spaced DPM-Solver steps, t=0.1 maps to λ≈1.08 and t=0.005 maps to λ≈3.56 in log-SNR space. The 10 steps span λ∈(−5, 4.26), so approximately 3 of the 10 steps fall inside the guidance window, not just 1.
 Regarding placement in the base class: agreed it should be documented and overridable. Fixed in the latest commit — _t_min and _t_max are now explicit class attributes on BaseGuidance with a docstring explaining the rationale, and subclasses can overridethem if a different window is needed for a specific guidance function, added in [c79812d](https://github.com/tier4/Diffusion-Planner/pull/22/commits/c79812def63b28dcda7a6d49384161543477d844)

**Review by SakodaShintaro — APPROVED** [review]
